src/generic/datavgen.cpp

```diff
@@ -2046,69 +2051,215 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
     return true;
 }
 
-void wxDataViewMainWindow::RemoveDropHint()
+void wxDataViewMainWindow::RemoveDropHint(bool refreshOnly/* = false*/)
 {
-    if (m_dropHint)
+    if (m_dropHint == 0)
+        return;
+
+    if (m_dropHint == 1)
+        RefreshRow(m_dropHintLine);
+    else if (m_dropHint == 2)
+        RefreshRows(m_dropHintLine == 0 ? 0 : m_dropHintLine - 1, m_dropHintLine);
+    else if (m_dropHint == 3)
+        RefreshRows(m_dropHintLine, std::max(m_dropHintLine + 1, GetRowCount() - 1));
+
+    if (!refreshOnly)
     {
-            m_dropHint = false;
-            RefreshRow( m_dropHintLine );
-            m_dropHintLine = (unsigned int) -1;
+        m_dropHint = 0;
+        m_dropHintIndentLevel = -1;
+        m_dropHintLine = (unsigned int) -1;
     }
 }
 
-wxDragResult wxDataViewMainWindow::OnDragOver( wxDataFormat format, wxCoord x,
-                                               wxCoord y, wxDragResult def )
+// Returns drop hint
+int wxDataViewMainWindow::GetDropItemInfo(const wxCoord x, const wxCoord y,
+                                          unsigned int* itemRow,
+                                          wxDataViewItem* item,
+                                          int* proposedDropIndex,
+                                          int* indentLevel)
 {
     int xx = x;
     int yy = y;
     m_owner->CalcUnscrolledPosition( xx, yy, &xx, &yy );
-    unsigned int row = GetLineAt( yy );
 
-    wxDataViewItem item;
+    unsigned int row = GetLineAt(yy);
+    *itemRow = row;     // [arg set]
+
+    if (row >= GetRowCount() || xx > GetEndOfLastCol())
+        return 0;
+
+    if (IsVirtualList())
+    {
+        *item = GetItemByRow(row);      // [arg set]
+
+        if (!item->IsOk())
+            return 0;
+    }
+    else    // TODO: DROP_INDEX_STYLE
+    {
+        wxDataViewTreeNode* node = GetTreeNodeByRow(row);
+        if (!node)
+            return 0;
+
+        *item = node->GetItem();
+
+        unsigned int itemStart = GetLineStart(row);
+        int itemHeight = GetLineHeight(row);
+
+        bool insertBefore = yy - itemStart < itemHeight*0.15;       // 15%, TODO: configurable
+        if (insertBefore)
+            // Can be treated as 'insertAfter" with the small difference:
+        {
+            node = GetTreeNodeByRow(row - 1);   // We need the node from the previous row
+
+            if (!node)
+                // Seems to be dropped in the root node
+            {
+                *indentLevel        = 0;                // [out]
+                *proposedDropIndex  = 0;                // [out]
+                *item               = wxDataViewItem(); // [out]
+
+                return 2;       // BEFORE
+            }
+
+        }
+
+        bool insertAfter = yy - itemStart > itemHeight*0.85;        // 15%, TODO: configurable
```

Comment:
15% seems a strange choice. Why not 50%?

Reply:
(100% - 2*15%) = 70% of the area goes to "drop inside".

50/50 should be checked after the first attempt of **OnDragOver** failure (when **OnDragOver** returns _wxDragNone_,_wxDragCancel_,_wxDragError_ after event handling)

This would be another PR ;)
In this case, **OnDrop** and **OnData** functions must use cached "drop data" from **OnDragOver** (like on macOS impl)!?